### ISSUE TITLE
rethink the entire (de)referencing concept

### ISSUE THREAD

**Issue by GlenDC** [g8os/blockstor#88]
Currently in master, we are (de)referencing deduped content, and deleting it when its reference counter is 0. This is all done in the same nbdserver process, which is a bit slow. On top of that we don't take into account the reference counting values, when copying metadata using the `copyvdisk` cmd CLI.

If wanted, the (de)reference counting as it is can be removed, by merging/rebasing the https://github.com/g8os/blockstor/tree/remove-content-referencing branch. This way, we can continue testing & developing, without suffering from issues, flowing out of the current (flawed) implementation (originally developed as a solution for issue #42).

The entire (de)referencing concept should be retaught, to see how we can better fit it in the entire cluster of blockstor projects. Ideally it should also not impact the speed of a performance-critical application, as it does now within the nbdserver.

@robvanmieghem suggested that we might be able to integrate this logic in the tlogserver instead, and do it based on the tlogs that nbdserver should send, once tlogserver is integrated (see: #40).

Tasks:

- [ ] provide option to delete the current (de)referencing logic (see https://github.com/g8os/blockstor/tree/remove-content-referencing);
- [ ] rethink the entire (de)referencing concept;
- [ ] finish #40 and #49;
- [ ] implement a new functional (de)referencing solution, which works in harmony with all the blockstor projects;

**Comment by GlenDC** [comment]
@robvanmieghem I'm not any longer (fully) convinced that tlog is a good place to de(reference) content:

1. A tlog agent has no idea if a volume is deduped or not, neither does the ardb.Backend where the tlogclient is used and integrated;
2. It feels a bit weird to have a totally separate module like this deal with the internals of the deduped storage (as it would have to if it wants to delete its content);

I don't have a better idea for now though, but wanted to throw out these intermediate thoughts of mine, to perhaps also find out why you initially thought the tlog module was a good new home for the de(referencing) of deduped content.

**Comment by iwanbk** [comment]
Do we have docs about the difference between deduped & non deduped mode?

**Comment by GlenDC** [comment]
> Do we have docs about the difference between deduped & non deduped mode?

@iwanbk: https://github.com/g8os/blockstor/blob/880d3de77c5b6bb9604713b55e9f2c3074164dfc/nbdserver/docs/backendStorage.md

**Comment by iwanbk** [comment]
> I'm not any longer (fully) convinced that tlog is a good place to de(reference) content:

@GlenDC if you have more context to share, i might be able to share my view about this.

**Comment by GlenDC** [comment]
@iwanbk so the entire (de)referencing only applies to deduped (as in de-duplicated) storage. As you hopefully know by now, the deduped storage (used only for `boot` vdisks, for now), stores the actual data as a separate value, using its blake2b hash as it its key. This hash (key) is then stored in the LBA (metadata), which is stored as a byte slice, 128 at a time (called a shard). This information can all be found in the documentation linked above.

The goal of this (de)referencing feature is that:

+ when zero LBA indices reference a certain hash (and thus content), that content should be deleted, as to not waste precious space;
+ each time content gets referenced (as in each time a hash is stored is some LBA index), that content is one more time referenced (a single vdisk could reference the content multiple times, and that should count as being multiple times);
+ the exception is full-zeroes content which doesn't get stored at all, and thus never referenced;
+ every time content is no longer referenced by an LBA index (because it is overwritten or deleted), it should dereference that content. And as noted above, when this means that no LBA index is any longer referencing this hash, the content is deleted forever;
  + BUT if that content was at some point referenced by AT LEAST 10 LBA Indices (or should it be vdisks???), that content is never deleted, even if at some point no one references it. I never understood why this exception, but this was stated as part of the feature by @robvanmieghem;

Originally this was implemented within the dedupedStorage code, as a goroutine, back when the storage interface still had an optional goroutine method, which was run via the `ardb.Backend` goroutine interface method. This was however slowing down the code quite a bit. On top of that, there was no support for this (de)referencing within the copyvdisk tool (now built-in the g8stor tool suite). So the new implementation will also need to be supported by all tools that interact with the deduped storage (thus all relevant commands in the g8stor tool suite).

At the same time that we removed the original implementation from the codebase (it's still in the git history though, the commit where we deleted that feature is here: ), @robvanmieghem suggested that we might be able to add this (de)referencing behavior to the tlog code. However I think it doesn't really belong there, and don't agree any longer with that idea, for the reasons stated in my earlier comment (https://github.com/g8os/blockstor/issues/88#issuecomment-301082635).

At the moment I still think the best place for (de)referencing content would still be the deduped storage. The actual reference counting is actually fairly cheap. So I think we could do the following:

+ (de)reference each time we modify content (this is cheap, and hopefully a neglectable cost);
+ run a swipe job, which checks for content with 0-referenced at the end of a vdisk (nbd) connection. This should in theory be OK, as we can do the actual heavy lifting purely on the ardb server using a Lua script; However it will be a stop-the-world moment, which will be a traceable moment both in the client (nbdserver) as well as on the server (ardb);
+ would still need to think on how to implement this "when-more-then-10-references-never-delete-content" exception though;

So yeah, I think that's my current improved design, based on the original implementation.

Besides that I don't have any fresh ideas. I certainly don't think it belongs in the tlog module, and that server is already suffering enough as it is, in terms of performance.

I hope that gives you enough context @iwanbk. All feedback, ideas and inspiration are welcome, as this is still an issue tagged for https://github.com/g8os/blockstor/milestone/6, together with the related issue #147, so time isn't on our side.

**Comment by iwanbk** [comment]
thanks @GlenDC for a very long explanation. I've read it but still can't help you.
I hope i get better view after i dig nbdserver code deeper.

**Comment by GlenDC** [comment]
**TLDR**: I think implementing deferred RC might be efficient enough that we could get away with it. It would require us to also resolve issue #351, and answer the open questions for the scrubber issue #436 (such as "do we need to really store a consumer list or can we simply store reference counting").

When I revived the original (naive) referencing counted (which was originally implemented a long time ago for about a week) in issue #299, I concluded that the naive RC is indeed too slow. Based on that I'm convinced that if we want it too work we'll have to built upon the research that has been done about RC. Based on my findings I was thinking about either going for weighted RC or deferred RC.

Currently we write all data (or delete, which is in a way also a write operation) immediately to the ARDB server. For the LBA metadata of deduped vdisks that is not the case, as we use a LRU cache for that.

I was thinking we could implement a dirty cache for both nondeduped and deduped disks. In it, we would only store dirty content (as in content that is either written, or deleted). This cache would than emptied and written to the ARDB server during a flush call. Such a cache would still need a limit, as I don't think we can simply rely that a flush will always come fast enough. And with 20 vdisks per NBDServer that are mounted maximum at the same time, that could become a lot of data in memory. So better to have a reasonable limit.

By going for such a "dirty" cache, as expressed above and originally proposed in issue #351, we can also implement the reference counting for deduped storage blocks in a deferred wait, and thus go for deferred RC. This is possible as the data updates would only ever happen together with the RC updates, and thus a scrubber can still do its work at any time it wants without being afraid it might remove data it shouldn't have removed.

The only open question would be how to keep this in sync with the LBA. For which I suppose the easy answer would be to not write any metadata until we flush our dirty cache. Is that the desirable answer is a different question.

**Comment by muhamadazmy** [comment]
I don't see how the `dirty` cache will help us with this situation because of the following points (correct me if i am wrong):

- Writing the block data and increment of the counter is not an atomic operation, although it doesn't have a direct impact on the performance anymore due to deferred flush operation.
- Even if we used a lua script on the ardb server side for setting and counter incr/decr it still not atomic because ardb is multi-threaded (unlike redis)
- Due to the previous 2 points, race conditions can happens in case 2 disks are sharing and updating the same block at the same time (i know this can be very low chance but still can happen)
- If something went wrong (inc/dec fails for a reason or another) this block can remain in ardb forever (if decr fails), while it's not referenced at all. Or even worse, a block can be deleted while it still referenced by another vdisk (in case of incr fails) there is no way to detect it. 

I believe a GC process that scans and detect un-referenced blocks might be a better approach. The problem would be how to implement such a process without degrading the performance of the ardb cluster.

**Comment by muhamadazmy** [comment]
I checked if ardb can grantee atomicity of operations, and it seems the only way to do so is to configure it to use a single thread, in that case TRANSACTIONS will be granted  to be atomic (also execution of lua scripts).

I am not sure about how ardb is configured in our setup, if we can (or already use) a single threaded setup, it will be possible that we do an atomic set of block data via a lua script.

In that case the block data and the ref counter is incremented (or decremented) atomically, the script can also do the delete operation of the block data if the counter hits zero.

On disk copy, another atomic check and increment script can be used to bump the ref counter by one without actual set of the block data, which should fail if the counter is not set (the block has been removed).

**Comment by GlenDC** [comment]
> Writing the block data and increment of the counter is not an atomic operation, although it doesn't have a direct impact on the performance anymore due to deferred flush operation.

Wrong. There is no deferred flush operation. The only thing that might not be deferred are the cached LBA sectors. But the data (blocks) itself is written immediately for both storage types. So if you wouldn't work with an intermediate cache (which is perhaps for the best) you'll have to also modify the reference each time. This is the naive reference counting which we tried at first, and it does have a significant impact on the performance, even when using a lua script.

> I believe a GC process that scans and detect un-referenced blocks might be a better approach. The problem would be how to implement such a process without degrading the performance of the ardb cluster.

That was already the plan. See #436. It wouldn't really be a process though, it would be more a deamon or process calling a new `zeroctl` command at interval timings. You however need to ensure that the mounted deduped vdisks in the nbdserver update the reference counts of the blocks it modifies. The relevant `zeroctl copy/delete/import/restore` commands also need to update these reference counts (this will be trickier to do and have an impact of the performance of those commands for sure).

> I checked if ardb can grantee atomicity of operations, and it seems the only way to do so is to configure it to use a single thread, in that case TRANSACTIONS will be granted to be atomic (also execution of lua scripts).

Regardless of the server type and configuration, it would be better if you can guarantee from a client side perspective (zeroctl and the nbdserver), that these operations are atomic. And most important, it really shouldn't have to impact performance of the nbdserver, as it already isn't as fast as we would ideally want.

**Comment by GlenDC** [comment]
> I don't see how the dirty cache will help us with this situation because of the following points (correct me if i am wrong).

It was just an idea of mine, perhaps it's a stupid one though. But it has two obvious benefits:

+ It would increase performance, by updating multiple data blocks and reference counts at once, at (not time-related) intervals;
+ It would also decrease the chance of getting data races, as you wouldn't be updating those reference counters every time. On top of that it means that even if someone just deleted a block we care about, we still have it in cache and will be writing it ourselves (again) when we need to evict the cache (due to space limitations) or because the flush command is called;

Caching is never easy though, so getting this right would be crucial. It would also make the vdisks much more vulnerable in case it would crash, as all the data in the cache would be lost. I think it's worth it though given that it can give us the benefits that i just mentioned. But perhaps you have a different perspective on it.

**Comment by muhamadazmy** [comment]
@GlenDC  there is a huge misunderstanding and i guess we will need to have a 1 to 1 meeting to discuss this.

>here is no deferred flush operation. The only thing that might not be deferred are the cached LBA sectors. But the data (blocks) itself is written immediately for both storage types. 

Yesterday i tried to implement an intermediate cache to buffer the written blocks before they are committed to storage, but then realised you actually don't want to defer the writing of the blocks, only the reference increments/decrements. Although this contradicts with #351!! So I am not sure what the requirements are exactly.

I still don't get how that is gonna help us accomplish the task in hand. As I mentioned before, to implement it the right way we need to grantee that the increments are done atomically with the block data writing.
There is no way this atomicity can be granted from client side, not even using 'MULTI' transaction support of redis/ardb unless ardb  is configured to be single threaded.

Now, for the cache problem, if we still gonna write the block directly to storage, and buffer it in the dirty cache until a flush or a buffer size limit is reached, wouldn't this cause the same block to be written twice to the storage (eventually), plus the reference operation, which are also not atomic.

I think discussing the issue and the implementation details requires a talk so we can settle on the right approach to fix this issue.
